feat: add support for service endpoints with location restrictions

matt-FFFFFF opened the pull request:
- Add service_endpoints_with_location variable to main module and subnet submodule
- Implement backward compatibility with existing service_endpoints variable
- Add validation to prevent simultaneous use of both variables
- Include comprehensive example demonstrating various location scenarios
- Maintain AVM compliance with separate attribute approach

Resolves #16

kewalaka commented:
nice work, this goes a long way to fix #17 from my perspective.  

The outputs should look something like this (noting the reference to the AVM spec).

```terraform
# Module owners should include the full resource via a 'resource' output
# https://azure.github.io/Azure-Verified-Modules/specs/terraform/#id-tffr2---category-outputs---additional-terraform-outputs
output "resource" {
  value       = azurerm_virtual_network.this
  description = "This is the full resource output for the virtual network resource."
}
```

Since I assume we want to encourage use of output through the 'resource' full output, I would recommend removing the others.

kewalaka commented:
I would suggest moving the subnet code to "main.subnet.tf" - similar to the child resource pattern used by Matt White in the KeyVault module.  More crucially, the child resources also need to support the Terraform interfaces, i.e. you need to be able to set role assignments on subnets too (I don't think subnets support locks).

kewalaka commented:
The AVM template has an updated ```terraform.tf``` - you need at least 1.3.0 to support the optional keyword, from the template:

```
terraform {
  required_version = ">= 1.3.0"
  required_providers {
    # TODO: Ensure all required providers are listed here.
    azurerm = {
      source  = "hashicorp/azurerm"
      version = ">= 3.71.0"
    }
    random = {
      source  = "hashicorp/random"
      version = ">= 3.5.0"
    }
  }
}
```

kewalaka commented:
The mandatory files check needs fixing, I would suggest updating the ./.github/actions, policies & workflows folder from the template to make sure this resource is aligned to the required checks - I can spot several differences.

For background - the docs-check had a bug that was fixed here, which might have been blocking you previously
https://github.com/Azure/terraform-azurerm-avm-template/issues/37

kewalaka commented:
the AVM template now ignores terraform lock files, on line 29, add:

```
# Ignore Terraform lock file
.terraform.lock.hcl
```

kewalaka commented:
in the locals.tf - I would personally use a single locals block, and you can remove the "TODO" & commented out code at the top of the file.

mbilalamjad reviewed (approved):
Approving for now to unblock, will review in detail and request a deeper discussion around some of the changes especially to files of the repo template.

matt-FFFFFF committed:
feat: add support for service endpoints with location restrictions

- Add service_endpoints_with_location variable to main module and subnet submodule
- Implement backward compatibility with existing service_endpoints variable
- Add validation to prevent simultaneous use of both variables
- Include comprehensive example demonstrating various location scenarios
- Maintain AVM compliance with separate attribute approach

Resolves #16

matt-FFFFFF committed:
chore: update to new gov

matt-FFFFFF committed:
use avm regions module

matt-FFFFFF committed:
chore: regions module again

matt-FFFFFF committed:
fix: bug with service endpoint location default

matt-FFFFFF committed:
chore: pre-commit

matt-FFFFFF committed:
fix: use ordered lists

matt-FFFFFF committed:
fix: cycle in validation

matt-FFFFFF committed:
fix: null iteration

matt-FFFFFF committed:
fix: function

matt-FFFFFF committed:
lists

matt-FFFFFF committed:
docs: update example with all service endpoints

jaredfholgate reviewed (approved):
LGTM
